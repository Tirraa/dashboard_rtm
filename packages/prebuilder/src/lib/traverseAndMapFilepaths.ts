import type { Stats } from 'fs';

import type { Filename, File, Path } from '../types/Metadatas';
import type { Thunks, Thunk } from '../types/LazyEvaluation';

type ArborescenceFragmentPromise = Promise<File[] | File>;

// https://github.com/vitest-dev/vitest/discussions/2484
const path = require('path');
const fs = require('fs/promises');

async function makeThunks(currentFolder: Path, __currentDeepPath: Path = currentFolder): Promise<Thunks<ArborescenceFragmentPromise>> {
  const files: Filename[] = await fs.readdir(currentFolder);
  const thunks: Thunks<ArborescenceFragmentPromise> = files.map((currentFilename: Filename) => {
    const maybeFilepath: Path = path.join(currentFolder, currentFilename);
    const thunk: Thunk<ArborescenceFragmentPromise> = async () => {
      const filepathStats: Stats = await fs.stat(maybeFilepath);
      return filepathStats.isDirectory()
        ? execute(await makeThunks(maybeFilepath, path.join(__currentDeepPath, currentFilename)))
        : ({ directory: __currentDeepPath, name: currentFilename } satisfies File);
    };
    return thunk;
  });

  return thunks;
}

const execute = async (thunks: Thunks<ArborescenceFragmentPromise>): Promise<File[]> => (await Promise.all(thunks.map((thunk) => thunk()))).flat();

const traverseAndMapFilepaths = async (rootFolder: Path): Promise<File[]> => execute(await makeThunks(rootFolder));

export default traverseAndMapFilepaths;
