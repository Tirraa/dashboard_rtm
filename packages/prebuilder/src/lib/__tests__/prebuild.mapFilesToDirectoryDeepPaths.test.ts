// eslint-disable-next-line import/no-extraneous-dependencies
import { describe, expect, it } from 'vitest';

import traverseAndMapFilepaths from '../traverseAndMapFilepaths';

const INVALID_PATH = './bless/the/draco/these/rounds/holy/sorry/bout/your/luck';

describe('traverseAndMapFilepaths', () => {
  it('should return an arborescence, given a valid RootFolder (full)', async () => {
    expect(await traverseAndMapFilepaths('./packages/prebuilder/src/lib/__tests__/fakeDirectory')).toMatchSnapshot();
  });

  it('should return an arborescence, given a valid RootFolder (partial)', async () => {
    expect(await traverseAndMapFilepaths('./packages/prebuilder/src/lib/__tests__/fakeDirectory/bar')).toMatchSnapshot();
  });

  it('should return an arborescence, given a valid RootFolder (partial 2)', async () => {
    expect(await traverseAndMapFilepaths('./packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz/foo')).toMatchSnapshot();
  });

  it('should throw ENOENT, given invalid path', async () => {
    expect.assertions(1);

    try {
      await traverseAndMapFilepaths(INVALID_PATH);
    } catch (e) {
      const interceptedError = e as Error;
      if ('code' in interceptedError) {
        expect(interceptedError.code).toBe('ENOENT');
      } else {
        throw new Error('Error code not found');
      }
    }
  });
});
