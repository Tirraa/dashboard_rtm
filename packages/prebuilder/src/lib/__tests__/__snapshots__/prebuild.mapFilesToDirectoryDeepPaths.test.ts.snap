// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`traverseAndMapFilepaths > should return an arborescence, given a valid RootFolder (full) 1`] = `
[
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz",
    "name": "file1",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz",
    "name": "file2",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz",
    "name": "file3",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz/foo",
    "name": "file1",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar",
    "name": "file1",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar",
    "name": "file2",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/baz/bar",
    "name": "file1",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/baz/bar/foo",
    "name": "file1",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/baz/bar/foo",
    "name": "file2",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/baz",
    "name": "file1",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/baz",
    "name": "file2",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/baz",
    "name": "file3",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/foo/baz/bar",
    "name": "file1",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/foo/baz/bar",
    "name": "file2",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/foo/baz",
    "name": "file1",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/foo",
    "name": "file1",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/foo",
    "name": "file2",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/foo",
    "name": "file3",
  },
]
`;

exports[`traverseAndMapFilepaths > should return an arborescence, given a valid RootFolder (partial 2) 1`] = `
[
  {
    "directory": "./packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz/foo",
    "name": "file1",
  },
]
`;

exports[`traverseAndMapFilepaths > should return an arborescence, given a valid RootFolder (partial) 1`] = `
[
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz",
    "name": "file1",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz",
    "name": "file2",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz",
    "name": "file3",
  },
  {
    "directory": "packages/prebuilder/src/lib/__tests__/fakeDirectory/bar/baz/foo",
    "name": "file1",
  },
  {
    "directory": "./packages/prebuilder/src/lib/__tests__/fakeDirectory/bar",
    "name": "file1",
  },
  {
    "directory": "./packages/prebuilder/src/lib/__tests__/fakeDirectory/bar",
    "name": "file2",
  },
]
`;
